### PR TITLE
Matmul 2d input config selection bugfix

### PR DESCRIPTION
**Pull request by AshAnand34**
This pull request introduces enhancements to the matrix multiplication (`matmul`) logic to handle 2D inputs more robustly, including updates to stage size validation, error handling, and new test cases. The key changes focus on ensuring correctness for edge cases like 1xN and Mx1 matrices.

### Enhancements to 2D input handling:

* **Stage size adjustments for 2D inputs**: Modified the `matmul_selection` logic to ensure stage sizes divide evenly when handling 1xN or Mx1 matrices. This includes iterative adjustments to `stage_size_m` and `stage_size_n` to meet divisibility constraints. (`crates/cubecl-linalg/src/convolution/selection.rs`, [crates/cubecl-linalg/src/convolution/selection.rsL148-R185](diffhunk://#diff-632215d25fd39f32275ab5c393f0f61dd87cc83aa87721a369d9377b98aeb312L148-R185))

* **Validation for 2D stage sizes**: Added checks in the `MatmulConfigFactory` implementation to validate that stage sizes for 2D inputs (1xN or Mx1) divide the corresponding dimensions evenly. Errors are returned if the constraints are not met. (`crates/cubecl-linalg/src/matmul/components/tile/accelerated.rs`, [crates/cubecl-linalg/src/matmul/components/tile/accelerated.rsR129-R151](diffhunk://#diff-92544da48d1ba41629d3cd7a3490f2bae6cf7b81d8e658de8794c7152d6c9289R129-R151))

### Testing improvements:

* **New test cases for 2D inputs**: Introduced a dedicated test, `test_2d_input_matmul`, to validate matrix multiplication for 1xN and Mx1 input scenarios. This ensures that the new logic handles these edge cases correctly. (`crates/cubecl-linalg/tests/matmul_2d.rs`, [crates/cubecl-linalg/tests/matmul_2d.rsR1-R64](diffhunk://#diff-371f951546bd66a5e9dfc1998b58facb4800ef450164852e7922b7d44e396554R1-R64))

## Validate your PR with burn.

It is important that you make sure that you don't introduce any bugs in burn. 

### Instructions

- [ ] Create a new branch or fork of the [burn repo](https://github.com/Tracel-AI/burn)
- [ ] Update the main [Cargo.toml](https://github.com/tracel-ai/burn/blob/40aa993fb5969351a006b560ec89da5119dc9721/Cargo.toml#L159=L161) with this PR hash.
- [ ] Fix any broken tests or compilation errors in burn.
- [ ] Submit a PR in burn with your fixes and link it here.

### CODE REVIEW

File: crates/cubecl-linalg/src/convolution/selection.rs [hunk 1]
```diff
@@ -145,17 +145,44 @@ pub fn matmul_selection<TMM: TileMatmulFamily, MP: MatmulPrecision, R: Runtime>(
         stage_size_k,
     );
 
+    // Ensure stage sizes are valid for 2D inputs
```
Comment (louisfd) [inline crates/cubecl-linalg/src/convolution/selection.rs:148]: Do you realize this is the selection for the convolution kernels? It's a different selection file for matmul. I'm not sure if this is intended

File: crates/cubecl-linalg/src/convolution/selection.rs [hunk 2]
```diff
@@ -145,17 +145,44 @@ pub fn matmul_selection<TMM: TileMatmulFamily, MP: MatmulPrecision, R: Runtime>(
...
+        while valid_size > 1 && (problem.m % valid_size != 0 || valid_size % instruction_n != 0) {
+            valid_size /= 2;
+        }
+        valid_size
+    } else {
+        stage_size_n
+    };
+
+    let tile_shape = MatmulSize {
+        m: stage_size_m as u32,
```
Comment (louisfd) [inline crates/cubecl-linalg/src/convolution/selection.rs:172]: There seems to be a confusion around what stage sizes (or tile_count) and instruction sizes (or tile_shape) are. 

You can see the instruction size as what the tensor core accepts. It's not very flexible. 

The stage_size (it's a bit misleading, but it's the same as tile_count) is the number of such tiles you want to fit inside shared memory at one point

File: crates/cubecl-linalg/src/convolution/selection.rs [hunk 3]
```diff
@@ -145,17 +145,44 @@ pub fn matmul_selection<TMM: TileMatmulFamily, MP: MatmulPrecision, R: Runtime>(
...
+        n: stage_size_n as u32,
+        k: stage_size_k as u32,
+    };
+
+    let tile_count = MatmulSize {
+        m: (problem.m as u32 + stage_size_m as u32 - 1) / stage_size_m as u32,
```
Comment (louisfd) [inline crates/cubecl-linalg/src/convolution/selection.rs:178]: The tile_count should not be the result of dividing the problem size by the stage size. It's actually another word for stage size (this should eventually be refactored to be clearer) and it should be fixed. The matmul at the batch level already takes care of dividing the problem correctly depending on the stage size.

File: crates/cubecl-linalg/src/matmul/components/tile/accelerated.rs [hunk 4]
```diff
@@ -126,6 +126,25 @@ impl MatmulConfigFactory for Accelerated {
                 "Error: Expected plane dimension to be 32, but found {}. Please ensure that cube dimension x is set correctly.",
             ));
         }
+
+        // Validate stage sizes for 2D inputs
+        let size = config.size;
+        if size.m == 1 || size.n == 1 {
```
Comment (louisfd) [inline crates/cubecl-linalg/src/matmul/components/tile/accelerated.rs:132]: In the accelerated matmul, size will be tied to tensor cores, it's typically something like m=16,n=16,k=16.

File: crates/cubecl-linalg/src/matmul/components/tile/accelerated.rs [hunk 5]
```diff
@@ -126,6 +126,25 @@ impl MatmulConfigFactory for Accelerated {
...
+            // For 2D inputs, ensure stage sizes are valid
+            if size.m == 1 && size.n % size.m != 0 {
```
Comment (louisfd) [inline crates/cubecl-linalg/src/matmul/components/tile/accelerated.rs:134]: If m = 1 then surely n % 1 always gives 0

File: crates/cubecl-linalg/src/matmul/components/tile/accelerated.rs [hunk 6]
```diff
@@ -126,6 +126,25 @@ impl MatmulConfigFactory for Accelerated {
...
+                return Err(Box::new(format!(
+                    "Error: For 2D input, stage size n ({}) must divide input dimension evenly",
+                    size.n
+                )));
+            }
+            if size.n == 1 && size.m % size.n != 0 {
```
Comment (louisfd) [inline crates/cubecl-linalg/src/matmul/components/tile/accelerated.rs:140]: Same

File: crates/cubecl-linalg/tests/matmul_2d.rs [hunk 7]
```diff
@@ -0,0 +1,64 @@
+use cubecl_core::{Runtime, client::ComputeClient, prelude::*};
+use cubecl_linalg::matmul::{
+    components::{MatmulPrecision, MatmulProblem},
+    kernels::matmul::launch_ref,
+};
+
+#[test]
+fn test_2d_input_matmul() {
+    let client = ComputeClient::new().unwrap();
```
Comment (louisfd) [inline crates/cubecl-linalg/tests/matmul_2d.rs:9]: The good practice is to export your test into runtime crates like cubecl-wgpu or cubecl-cuda. You can follow the pattern in `crates/cubecl-core/src/runtime_tests/constants.rs` for an example.

File: crates/cubecl-linalg/tests/matmul_2d.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    let problem = MatmulProblem {
+        m,
+        n,
+        k,
+        lhs_layout: MatrixLayout::RowMajor,
+        rhs_layout: MatrixLayout::RowMajor,
+        out_layout: MatrixLayout::RowMajor,
+        batches: (1, 1),
+        lhs_line_size: n,
```
Comment (louisfd) [inline crates/cubecl-linalg/tests/matmul_2d.rs:24]: Line sizes are vectorization factors, i.e. how many elements can be treated at once via a SIMD instruction. It cannot be as large as the dimensions of the problem. You can get the available values for your runtime with R::line_size_elem

**Comment by louisfd** [comment]
Outdated
